Replicated event sourcing

chbatey opened the pull request:
Merge of the feature branch back into master

It has just been rebased 

https://github.com/akka/akka/issues/29217

chbatey committed:
Initial prototype of active active event sourcing (#29265)

chbatey committed:
ActiveActive: Events with metadata and events by persistence id for  (#29287)

chbatey committed:
Alternative approach for storing metadata (#29309)

johanandren committed:
Speculative replication - step 1 (#29289)

* Step 1: general event-publishing-to-topic feature

* Step 2: an actor subscribes to the topic and forwards events to the sharded replicas

* Another half piece of the puzzle, receive the PublishedEvent in the ESB internals

* Stash published events while replaying rather than drop

* Publish on the event stream instead of a topic

* Active active actor receiving a published event

* Some smaller changes

 * Public API for published event
 * Better name for the sharding component

* Naive test for the Active Active Sharding Replication

* Java API for ActiveActiveShardingReplication

* Spelling

* Use ShardingEnvelope for publishing the event across sharding

* Fast forwarding filter stage

* Move test to testkit, enable the see-event-twice test (fails)

* Use persistence testkit journal

* Various smaller review feedback things

* Trying to figure out why duplicate event write test fails

* Missing unstash after processing published event

Co-authored-by: Christopher Batey <christopher.batey@gmail.com>

patriknw committed:
better timestamp format for active-active (#29329)

patriknw committed:
message adapter in ActiveActiveShardingDirectReplication (#29328)

* I don't think it's worth cheating with Any and narrow just to save
  one or two allocations given how many other things that are needed for
  sending the messages

johanandren committed:
Support persistAll for active active (#29337)

johanandren committed:
Test coverage for active active published events after replay (#29335)

* Test coverage for published events after recovery
* Keep track of origin seqnrs when replaying

chbatey committed:
Introduce version vectors for replicated events (#29332)

* ActiveActive: Events with metadata and events by persistence id for  (#29287)
* Introduce version vectors for replicated events

* Set concurrent on recovery

* Group together AA related fields in a published event

johanandren committed:
Active active per replica journal selection (#29350)

* Support for having multiple isolated testkit journals and corresponding read journals in persistence testkit
* Read journal selection per active-active replica 
* a dedicated ReplicaId type to avoid stringly typed mismatches

chbatey committed:
Initial docs for active active (#29386)

chbatey committed:
Metadata for snapshots for active active (#29362)

chbatey committed:
Port CRDTs from multi dc (#29372)

* Metadata for snapshots for active active

* Port CRDTs from multi dc

* Review feedback

johanandren committed:
Active active java apis #29266

chbatey committed:
Docs and example for triggers (#29396)

johanandren committed:
Active active sharding (#29364)

johanandren committed:
Make sure to gracefully stop replication streams when aa actor stops #29406

johanandren committed:
Incorrect assumption about event ordering in test fixed (#29411)

johanandren committed:
EventPublishing API for Java (#29413)

johanandren committed:
Don't log user event toString in active active #29393 (#29416)

johanandren committed:
ActiveActiveSharding make it possible to use with custom envelope (#29417)

johanandren committed:
Active active docs mention plugin support (#29418)

* Mention specific support for active active is needed in plugins
* Align metadata name journal vs snapshot

chbatey committed:
Detect illegal access to context (#29431)

johanandren committed:
Active active journal support changes (#29421)

* Add missing serializers (and rename classes because metadata is a word)
* Add test instances for journal and snapshot support
* Method to add metadata to existing snapshotmetadata
* Better error message if replicated stream does not have metadata
* Snapshot store tck support for metadata
* Docs for tck coverage of AA metadata

chbatey committed:
PersistentRepr serializer support for metadata (#29434)

And fixing the warnings for the 2.13 build, not sure why PR
validation passes for these

The write side of the JDBC plugin works now as it
persists the PR

johanandren committed:
Rename active active to replicated event sourcing (#29435)

* Move active active internals into the internal package

* Renaming active active to Replicated Event Sourcing

* Rename of Active Active to Replicated Event Sourcing

* Revert changes to testkit

* Java test formatting...

* Gave up on getting apidoc to link to EventSourcedBehavior and made code snippets of it

Co-authored-by: Christopher Batey <christopher.batey@gmail.com>

johanandren committed:
Test case for three replicated event sourced entities added (#29438)

johanandren committed:
Document hot standby for replicated event sourcing (#29439)

chbatey committed:
Restart replication stream from correct seen seqNr (#29436)

chbatey committed:
Java auction example (#29443)

johanandren committed:
Tagging for replicated event sourcing (#29442)

* Tagging for replicated event sourcing
* Docs improvements
* Support for currentEventsByTag in persistence testkit and test coverage for tags in replicated

patriknw committed:
Minor adjustments to replicated es docs (#29450)

* Minor adjustments to replicated es docs

* and a few aa leftovers

* link to more examples

patriknw committed:
DRY ReplicatedEventSourcedBehavior.apply (#29454)

* and change signature of withEventPublishing and withDirectReplication

chbatey committed:
Embed akka serialization inside jackson serialization (#29455)

* Embed akka serialization inside jackson serialization

Allows re-use of existing serializers when embedding the types in
Jackson

* Fix false dead code warning

* Move marker trait and add docs

* Remove module and document adding annotations

* Review feedback

chbatey committed:
Rename replicated settings (#29460)

patriknw committed:
a few more things when reviewing replicated event sourcing (#29461)

chbatey committed:
Add entity type name for replicated entity (#29462)

* Add entity type name for replicated entity

* Review feedback

* Unused import

patriknw committed:
Polish Auction example

* adjust the style in the Scala example
  * AuctionEntity class instead of Setup class that is passed around
* add timer in recovery completed

patriknw committed:
LwwTime example

* show event and command handlers of the blog example in docs
* add blog example for Java

patriknw committed:
ORSet example, MovieWatchList

patriknw committed:
Counter example, shopping cart

chbatey committed:
Replicated Sharding improvements (#29483)

* WIP

* Finishing touches to sharding updates

* Review feedback

chbatey committed:
Turn on direct replication by default (#29493)

chbatey committed:
Make ShardingDirectReplication private (#29492)

* Make ShardingDirectReplication private

And provider user API to provide ReplicaId so that local messages don't
need to be forwarded to sharding to just be dropped

* Update docs to say direct replication is on my default

* Doh

chbatey committed:
Test coverage and fixes for direct replication (#29491)

* Test coverage and fixes for direct replication

* Serializer test and review feedback

* use testkit serialization testkit

chbatey committed:
Ignore java warnins in persistence typed and mima

Warnings are from protobuf

chbatey committed:
2.13 compilation

patriknw commented on akka-cluster/src/main/scala/akka/cluster/Member.scala:
```diff
@@ -41,9 +41,9 @@ class Member private[cluster] (
   override def toString =
     if (dataCenter == ClusterSettings.DefaultDataCenter)
-      s"Member(address = $address, status = $status)"
+      s"Member(address = $address, status = $status, roles = $roles)"
```
prettier
```
[${roles.mkString(", ")}]
```

patriknw commented on akka-docs/src/main/paradox/project/examples.md:
```diff
@@ -61,11 +61,7 @@ from the events, or publish the events to other services.
-
-Illustrates how to use Lightbend's [Multi-DC Persistence](https://doc.akka.io/docs/akka-enhancements/current/persistence-dc/index.html)
-with active-active persistent entities across data centers.
+This commercial feature has now been superseeded by @ref[Replicated Event Sourcing](../typed/replicated-eventsourcing.md)
```
this can be updated again when the sample has been updated

patriknw reviewed (approved):
LGTM, I have reviewed this in previous PRs and dev branch. Great work @johanandren and @chbatey !

patriknw commented:
Have we run this in a repeat jenkins job? Just to ensure that there are no surprises in multi-node tests.

chbatey commented:
Not yet but i will before merging

raboof commented on akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ReplicatedShardingExtensionImpl.scala:
```diff
@@ -0,0 +1,106 @@
+          replicaSettings.replicaId,
+          replicaSettings.entity.typeKey,
+          regionOrProxy,
+          replicaSettings.entity.dataCenter)
```
wonky formatting, but I guess that's scalafmt

raboof reviewed (commented):
Some small docs comments, what I've seen looks good (but haven't looked at everything)

chbatey commented:
Given that we don't plan to squash this branch i'll keep pushing feedback into a single commit after we have the other PR into this branch merged

chbatey commented:
Running on https://jenkins.akka.io:8498/job/legacy-akka-multi-node-repeat4/

chbatey committed:
Feedback from creating replicated entity sample (#29510)

* Fix javadsl and remove shard regions from ReplicatedSharding

* Simplyfy sharding API for replicated event sourcing

As the ShardRegion access has been removed then we will only initially
support Entity's with ShardingEnvelope meaning we can remove the type
param.

Also provide convenience constructors for running a replica on a role
and a replica in each DC

* Compile

* Review feedback

* feedback

chbatey committed:
Final review feedback

johanandren reviewed (approved):
Looked at the individual PRs done while I was vacationing and I think everything LGTM

patriknw commented:
Let's fix the conflict and then merge this.

patriknw committed:
Merge branch 'master' into feature-active-active-event-sourcing

patriknw commented:
I fixed the conflict. Let's have a few successful runs of multi-node tests before merging. Running at https://jenkins.akka.io:8498/job/legacy-akka-multi-node-repeat4/

patriknw committed:
Merge branch 'master' into feature-active-active-event-sourcing

patriknw committed:
fix compilation errors

patriknw commented:
Several successful runs in the multi-node tests. Some failures in unrelated cluster.StressSpec https://github.com/akka/akka/issues/29512

patriknw commented:
Merged 🎉 
Great work @johanandren and @chbatey !
